src/graph.rs

```diff
@@ -2555,6 +2445,230 @@ impl<'a, 'graph> Builder<'a, 'graph> {
+      npm_specifiers.push(specifier.clone());
+      if seen_specifiers.insert(specifier.clone()) {
+        let npm_infos: &mut Vec<_> = pending_npm_by_name
+          .entry(npm_ref.req.name.clone())
+          .or_default();
+        npm_infos.push((specifier, npm_ref, maybe_range));
+      }
+    }
+
+    Self {
+      npm_resolver,
+      pending_info: NpmSpecifierBuildPendingInfo::with_capacity(capacity),
+      pending_npm_by_name,
+      npm_specifiers,
+    }
+  }
+
+  async fn resolve(
+    &mut self,
+    mut pending_npm_registry_info_loads: PendingNpmRegistryInfoLoadFutures,
+  ) {
+    let mut previously_restarted = false;
+    while let Some((package_name, result)) =
+      pending_npm_registry_info_loads.next().await
+    {
+      let infos = self.pending_npm_by_name.get(&package_name).unwrap();
+      if let Err(err) = result {
+        for (specifier, _, maybe_range) in infos {
+          // load failure
+          self.pending_info.module_slots.insert(
+            specifier.clone(),
+            ModuleSlot::Err(ModuleGraphError::ModuleError(
+              ModuleError::LoadingErr(
+                specifier.clone(),
+                maybe_range.clone(),
+                err.clone(),
+              ),
+            )),
+          );
+        }
+      } else {
+        for (specifier, npm_ref, maybe_range) in infos {
+          if let Some(npm_resolver) = &self.npm_resolver {
+            let resolution = npm_resolver.resolve_npm(&npm_ref.req);
+            match resolution {
+              NpmPackageReqResolution::Ok(pkg_id) => {
+                self
+                  .pending_info
+                  .specifier_resolutions
+                  .insert(specifier.clone(), pkg_id.clone());
+                let pkg_id_ref = NpmPackageNvReference {
+                  nv: pkg_id,
+                  sub_path: npm_ref.sub_path.clone(),
+                };
+                let resolved_specifier = pkg_id_ref.as_specifier();
+                if resolved_specifier != *specifier {
+                  self
+                    .pending_info
+                    .redirects
+                    .insert(specifier.clone(), resolved_specifier.clone());
+                }
+                self.pending_info.module_slots.insert(
+                  resolved_specifier.clone(),
+                  ModuleSlot::Module(Module::Npm(NpmModule {
+                    specifier: resolved_specifier,
+                    nv_reference: pkg_id_ref,
+                  })),
+                );
+              }
+              NpmPackageReqResolution::ReloadRegistryInfo(_)
+                if !previously_restarted =>
+              {
+                // the implementer should ideally never return this more than once,
+                // but in case they do, have this safety
+                previously_restarted = true;
+
+                // clear the current pending information and restart from scratch
+                pending_npm_registry_info_loads.clear();
```

Comment:
Actually, just clearing out these futures is maybe a little sketch to do if the code isn’t made to handle it? Maybe these should be drained and awaited then the implementor could implement  cancellation for the requests? It’s more work though. Maybe it should just be documented instead that deno_graph will do this?

Reply:
Won't dropping these future just cancel them?

Reply:
It does, but at the same time some subtle bugs can occur if someone doesn't expect a future to not continue at a certain point. I'm just going to update the doc string.